### PR TITLE
Enforce HTTPS update

### LINKED ISSUE #6792: Surface info for HTTPS Redirection Middleware
See discussion: https://github.com/aspnet/Home/issues/3176
Announcement: https://github.com/aspnet/Announcements/issues/301
Announcement discussion: https://github.com/aspnet/BasicMiddleware/issues/329

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #6792 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/security/enforcing-ssl?view=aspnetcore-1.0&branch=pr-en-us-6793)

One thing right off the bat. The third entry is *terribad* ...

> To specify an HTTPS port:
>
> * Set [HttpsRedirectionOptions.HttpsPort](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.httpsport). See the preceding example that shows how to set the port to 5001.
> * Set the `HTTPS_PORT` environment variable.
> * The middleware uses the port or ports defined by [IServerAddressesFeature.Addresses](/dotnet/api/microsoft.aspnetcore.hosting.server.features.iserveraddressesfeature.addresses).

What should it say about port discovery from the `IServerAddressesFeature.Addresses`?

cc/ @k3davis

### CODE REVIEW

File: aspnetcore/fundamentals/middleware/index.md [hunk 1]
```diff
@@ -192,13 +192,15 @@ ASP.NET Core ships with the following middleware components, as well as a descri
 | [CORS](xref:security/cors) | Configures Cross-Origin Resource Sharing. | Before components that use CORS. |
 | [Diagnostics](xref:fundamentals/error-handling) | Configures diagnostics. | Before components that generate errors. |
 | [ForwardedHeaders/HttpOverrides](/dotnet/api/microsoft.aspnetcore.builder.forwardedheadersextensions) | Forwards proxied headers onto the current request. | Before components that consume the updated fields (examples: Scheme, Host, ClientIP, Method). |
+| [HTTPS Redirection Middleware](xref:security/enforcing-ssl#require-https) | Redirect all HTTP requests to HTTPS (ASP.NET Core 2.1 or later). | Before components that consume the URL. |
+| [HTTP Strict Transport Security (HSTS) Middleware](xref:security/enforcing-ssl#http-strict-transport-security-protocol-hsts) | Security enhancement middleware (ASP.NET Core 2.1 or later). | Before components that consume the URL. When called with HTTPS Redirection Middleware, call HSTS Middleware before [UseHttpsRedirection](/dotnet/api/microsoft.aspnetcore.builder.httpspolicybuilderextensions.usehttpsredirection). |
```
Comment (Tratcher) [inline aspnetcore/fundamentals/middleware/index.md]: HSTS only adds response headers for requests that are already HTTPS. It is not order sensitive with UseHttpsRedirection. It should go after any request modifications (ForwardedHeaders, rewrite) and before any responses are sent.

File: aspnetcore/security/enforcing-ssl.md [hunk 2]
```diff
@@ -10,45 +10,56 @@ ms.technology: aspnet
...
+> * Close the connection with status code 400 (Bad Request) and not serve the request.
 
 <a name="require"></a>
 ## Require HTTPS
 
 ::: moniker range=">= aspnetcore-2.1"
-We recommend all ASP.NET Core web apps call `UseHttpsRedirection` to redirect all HTTP requests to HTTPS. If `UseHsts` is called in the app, it must be called before `UseHttpsRedirection`.
 
-The following code calls `UseHttpsRedirection` in the `Startup` class:
+We recommend all ASP.NET Core web apps call HTTPS Redirection Middleware ([UseHttpsRedirection](/dotnet/api/microsoft.aspnetcore.builder.httpspolicybuilderextensions.usehttpsredirection)) to redirect all HTTP requests to HTTPS. If HTTP Strict Transport Security (HSTS) Middleware ([UseHsts](/dotnet/api/microsoft.aspnetcore.builder.hstsbuilderextensions.usehsts)) is called in the app, it must be called before `UseHttpsRedirection`.
```
Comment (Tratcher) [inline aspnetcore/security/enforcing-ssl.md]: No ordering restriction here.

File: aspnetcore/security/enforcing-ssl.md [hunk 3]
```diff
@@ -10,45 +10,56 @@ ms.technology: aspnet
...
 
-[!code-csharp[sample](enforcing-ssl/sample/Startup.cs?name=snippet2&highlight=14-99)]
+[!code-csharp[](enforcing-ssl/sample/Startup.cs?name=snippet2&highlight=14-99)]
 
-* Sets `RedirectStatusCode`.
+* Sets [HttpsRedirectionOptions.RedirectStatusCode](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.redirectstatuscode).
 * Sets the HTTPS port to 5001.
 
-::: moniker-end
+To specify an HTTPS port:
```
Comment (Tratcher) [inline aspnetcore/security/enforcing-ssl.md]: Let's cover the automatic mechanisms before we talk about setting it manually.

- If you run Kestrel or HttpSys directly with HTTPS endpoints then the middleware can discover the ports via IServerAddressesFeature (so long as there's only one port).
- If you run behind a reverse proxy (including IIS/Express) then IServerAddressesFeature isn't available and the port must be configured.
- Visual Studio with set ASPNETCORE_HTTPS_PORT if IIS Express has HTTPS enabled.

You can configure it manually via:
- ASPNETCORE_HTTPS_PORT env var
- http_port config key (.json, command line, etc.)
- Set HttpsRedirectionOptions.HttpsPort in code.

Comment (guardrex) [reply aspnetcore/security/enforcing-ssl.md]: Perfect ... thanks. I'll get that in here.

**Comment by guardrex** [comment]
Try now. Drafted this ...

The following mechanisms set the port automatically:

* The middleware can discover the ports via [IServerAddressesFeature](/dotnet/api/microsoft.aspnetcore.hosting.server.features.iserveraddressesfeature) when the following conditions apply:
  - Kestrel or HTTP.sys is used directly with HTTPS endpoints.
  - Only **one port** is used by the app.
* Visual Studio is used with:
  - The `ASPNETCORE_HTTPS_PORT` environment variable set.
  - IIS Express has HTTPS enabled.

> [!NOTE]
> When an app is run behind a reverse proxy (for example, IIS, IIS Express), `IServerAddressesFeature` isn't available. The port must be manually configured.

The port is configured manually via the following approaches:

* The `ASPNETCORE_HTTPS_PORT` environment variable is set.
* The `http_port` host configuration key is set (for example, via *hostsettings.json* or a command line argument).
* The [HttpsRedirectionOptions.HttpsPort](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.httpsport) is set. See the preceding example that shows how to set the port to 5001.

If no port is set:

* Requests aren't redirected.
* The middleware logs a warning.
* The middleware is disabled.

... One too many "vias" tho. I'll swap the middle one with "with."

**Comment by jkotalik** [comment]
> launchSettings.json sets the ASPNETCORE_HTTPS_PORT environment variable.

Also the ASPNETCORE_URLS can set the url, which will be populated in the IServerAddressFeature.

Other than that, LTGM.

**Comment by guardrex** [comment]
I'm going to have to refactor this a bit. Stand-by.

**Comment by guardrex** [comment]
> Also the ASPNETCORE_URLS can set the url, which will be populated in the IServerAddressFeature.

I'll put that in the manual section.

I wonder about this ...

*launchSettings.json* sets the `ASPNETCORE_HTTPS_PORT` environment variable

For IIS Express, isn't the port set by `sslPort` in *lanuchSettings.json*?

**Comment by guardrex** [comment]
I'm inclined to say ...

> *launchSettings.json* sets the `sslPort` for IIS Express.

**Comment by guardrex** [comment]
The following mechanisms set the port automatically:

* The middleware can discover the ports via [IServerAddressesFeature](/dotnet/api/microsoft.aspnetcore.hosting.server.features.iserveraddressesfeature) when the following conditions apply:
  - Kestrel or HTTP.sys is used directly with HTTPS endpoints (also applies to running the app with Visual Studio Code's debugger).
  - Only **one port** is used by the app.
* Visual Studio is used:
  - IIS Express has HTTPS enabled.
  - *launchSettings.json* sets the `sslPort` for IIS Express.

> [!NOTE]
> When an app is run behind a reverse proxy (for example, IIS, IIS Express), `IServerAddressesFeature` isn't available. The port must be manually configured.

The port is configured manually using any of the following approaches:

* The `ASPNETCORE_HTTPS_PORT` environment variable is set.
* The URL is set with the `ASPNETCORE_URLS` environment variable.
* The `http_port` host configuration key is set (for example, via *hostsettings.json* or a command line argument).
* The [HttpsRedirectionOptions.HttpsPort](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.httpsport) is set. See the preceding example that shows how to set the port to 5001.

File: aspnetcore/security/enforcing-ssl.md [hunk 4]
```diff
@@ -10,45 +10,68 @@ ms.technology: aspnet
...
+[!code-csharp[](enforcing-ssl/sample/Startup.cs?name=snippet1&highlight=13)]
 
-The following code:
+The following code calls [AddHttpsRedirection](/dotnet/api/microsoft.aspnetcore.builder.httpsredirectionservicesextensions.addhttpsredirection) to configure middleware options:
 
-[!code-csharp[sample](enforcing-ssl/sample/Startup.cs?name=snippet2&highlight=14-99)]
+[!code-csharp[](enforcing-ssl/sample/Startup.cs?name=snippet2&highlight=14-99)]
 
-* Sets `RedirectStatusCode`.
+* Sets [HttpsRedirectionOptions.RedirectStatusCode](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.redirectstatuscode).
```
Comment (Rick-Anderson) [inline aspnetcore/security/enforcing-ssl.md]: Do we want to pre-pend this line with 
The preceding highlighted code:

File: aspnetcore/security/enforcing-ssl.md [hunk 5]
```diff
@@ -10,45 +10,68 @@ ms.technology: aspnet
...
+
+* The middleware can discover the ports via [IServerAddressesFeature](/dotnet/api/microsoft.aspnetcore.hosting.server.features.iserveraddressesfeature) when the following conditions apply:
+  - Kestrel or HTTP.sys is used directly with HTTPS endpoints (also applies to running the app with Visual Studio Code's debugger).
+  - Only **one port** is used by the app.
+* Visual Studio is used:
+  - *launchSettings.json* sets the `ASPNETCORE_HTTPS_PORT` environment variable.
+  - IIS Express has HTTPS enabled.
+
+> [!NOTE]
+> When an app is run behind a reverse proxy (for example, IIS, IIS Express), `IServerAddressesFeature` isn't available. The port must be manually configured.
```
Comment (Rick-Anderson) [inline aspnetcore/security/enforcing-ssl.md]: Can you add something like
When the port is not set,  Requests aren't redirected.

I know you say it below, maybe you don't need it here.

File: aspnetcore/security/enforcing-ssl.md [hunk 6]
```diff
@@ -10,45 +10,68 @@ ms.technology: aspnet
...
+
+The port is configured manually using any of the following approaches:
```
Comment (Rick-Anderson) [inline aspnetcore/security/enforcing-ssl.md]: The port  can be configured by: 

* Setting the `ASPNETCORE_HTTPS_PORT` environment variable.

**Comment by jkotalik** [comment]
> launchSettings.json sets the sslPort for IIS Express.

Just talked to @javiercn, this is correct. ASPNETCORE_HTTPS_PORT is only used by kestrel.

**Comment by guardrex** [comment]
@Rick-Anderson How about this ...

The preceding highlighted code:

* Sets [HttpsRedirectionOptions.RedirectStatusCode](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.redirectstatuscode).
* Sets the HTTPS port to 5001.

The following mechanisms set the port automatically:

* The middleware can discover the ports via [IServerAddressesFeature](/dotnet/api/microsoft.aspnetcore.hosting.server.features.iserveraddressesfeature) when the following conditions apply:
  - Kestrel or HTTP.sys is used directly with HTTPS endpoints (also applies to running the app with Visual Studio Code's debugger).
  - Only **one port** is used by the app.
* Visual Studio is used:
  - IIS Express has HTTPS enabled.
  - *launchSettings.json* sets the `sslPort` for IIS Express.

> [!NOTE]
> When an app is run behind a reverse proxy (for example, IIS, IIS Express), `IServerAddressesFeature` isn't available. The port must be manually configured.

The port can be configured by setting the:

* `ASPNETCORE_HTTPS_PORT` environment variable.
* URL with the `ASPNETCORE_URLS` environment variable.
* `http_port` host configuration key (for example, via *hostsettings.json* or a command line argument).
* [HttpsRedirectionOptions.HttpsPort](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.httpsport). See the preceding example that shows how to set the port to 5001.

If no port is set:

* Requests aren't redirected.
* The middleware logs a warning.
* The middleware is disabled.

**Comment by jkotalik** [comment]
> Only one port is used by the app.

Only one HTTPS port is used by the app.

May people specify a HTTP and HTTPS port in ASPNETCORE_URLS.

**Comment by guardrex** [comment]
Anything else @jkotalik?

@Rick-Anderson U good on this one?

File: aspnetcore/security/enforcing-ssl.md [hunk 7]
```diff
@@ -10,45 +10,71 @@ ms.technology: aspnet
...
+  - IIS Express has HTTPS enabled.
+  - *launchSettings.json* sets the `sslPort` for IIS Express.
+
+> [!NOTE]
+> When an app is run behind a reverse proxy (for example, IIS, IIS Express), `IServerAddressesFeature` isn't available. The port must be manually configured. When the port isn't set, requests aren't redirected.
+
+The port can be configured by setting the:
+
+* `ASPNETCORE_HTTPS_PORT` environment variable.
+* URL with the `ASPNETCORE_URLS` environment variable.
```
Comment (Tratcher) [inline aspnetcore/security/enforcing-ssl.md]: Separate ASPNETCORE_URLS (e.g. move it to the bottom of the list or a independent note). This configures the server and then the middleware can indirectly discover the https port via IServerAddressesFeature.

File: aspnetcore/security/enforcing-ssl.md [hunk 8]
```diff
@@ -10,45 +10,71 @@ ms.technology: aspnet
...
+* `http_port` host configuration key (for example, via *hostsettings.json* or a command line argument).
+* [HttpsRedirectionOptions.HttpsPort](/dotnet/api/microsoft.aspnetcore.httpspolicy.httpsredirectionoptions.httpsport). See the preceding example that shows how to set the port to 5001.
 
+If no port is set:
+
+* Requests aren't redirected.
+* The middleware logs a warning.
+* The middleware is disabled.
```
Comment (Tratcher) [inline aspnetcore/security/enforcing-ssl.md]: "Requests aren't redirected." and "The middleware is disabled." are redundant. The first one's better.

**Comment by k3davis** [comment]
Thanks for your time/efforts, @guardrex , this makes the situation much more clear!

**Pull request merged**